### ISSUE TITLE
Bug since WordPress 4.6 [with fix]

### ISSUE THREAD

**Issue by MichaelSmi** [YahnisElsts/plugin-update-checker#66]
I'm usind your code to update a custom plugin not in public WordPress repo. Since update to 4.6 every upgrade fails with curl error. This is caused by malformed url.

Here is the fix:
in https://github.com/YahnisElsts/plugin-update-checker/blob/master/plugin-update-checker.php#L1208
change:
`$update->package = $this->download_url;`
to:
`$update->package = preg_replace('/([^:])(\/{2,})/', '$1/', $this->download_url);`

**Comment by YahnisElsts** [comment]
Could you provide a bit more information? What does the malformed URL look like?

Normally this kind of fix should not be necessary because the update checker just uses the provided download URL as-is. If the URL is incorrect, that's usually a configuration issue.

**Comment by MichaelSmi** [comment]
The bug was reported on WordPress Issue tracker: https://core.trac.wordpress.org/ticket/37733
They want to normalize urls before getting the upgrade package. In 4.6.1 to support all non strictly programmed plugins:
> I feel that there are enough plugins affected by incorrect url concatenation to fix this.

The malformed url looks like follows: `http://example.com//?action=get_metadata&slug=plugin_slug`
But the url is not configured that way! There is no double forward slash in the config url.

**Comment by YahnisElsts** [comment]
Thank you for the additional information. It looks like that's a problem in the update server, so it's the server that should be fixed, not the update checker. The server should output a correct download URL (i.e. no double slashes) and the update checker shouldn't *need* to normalize it.

Are you using the default `index.php` for `wp-update-server` or did you customize it in any way, like providing your own `$serverUrl` or subclassing `Wpup_UpdateServer`? Please post any relevant customizations. I'll see if I can reproduce the issue.

**Comment by MichaelSmi** [comment]
I didn't change anything in the wp-update-server. I only send more custom url parameters to the server to parse them from the log file externaly on the server.

**Comment by YahnisElsts** [comment]
Is it a Windows server? Try replacing the `guessServerUrl` method in `/includes/UpdateServer.php` with this version and see if it fixes the URL:

```php
public static function guessServerUrl() {
	$serverUrl = ( self::isSsl() ? 'https' : 'http' );
	$serverUrl .= '://' . $_SERVER['HTTP_HOST'];
	$path = $_SERVER['SCRIPT_NAME'];

	if ( basename($path) === 'index.php' ) {
		$path = $dir = dirname($path);
		if ( DIRECTORY_SEPARATOR === '/' ) {
			// Fix Windows
			$path = str_replace('\\', '/', $dir);
		}
		//Make sure there's a trailing slash.
		if ( substr($path, -1) !== '/' ) {
			$path .= '/';
		}
	}

	$serverUrl .= $path;
	return $serverUrl;
}
```

**Comment by bnecreative** [comment]
Had the same issue and this fixed it for me. I'm on a linux server as a side note.

**Comment by flowdee** [comment]
same here

**Comment by YahnisElsts** [comment]
Which fix - the one posted by Michael, or the patch for `guessServerUrl`?

**Comment by flowdee** [comment]
Used your guessServerUrl patch and it works fine now

**Comment by bnecreative** [comment]
The guessServerUrl method.

**Comment by YahnisElsts** [comment]
All right, thank you for testing the patch. The bug has been fixed in wp-update-server:
https://github.com/YahnisElsts/wp-update-server/commit/fa70d438945120d030bcfd368ba70e9ad75704bb
